motionplan/constraint.go

```diff
@@ -132,41 +132,79 @@ func (c *constraintHandler) CheckConstraints(cInput *ConstraintInput) (bool, flo
 // if reportDistances is false, this check will be done as fast as possible, if true maximum information will be available for debugging.
 func newSelfCollisionConstraint(
 	frame referenceframe.Frame,
-	observationInput map[string][]referenceframe.Input,
-	collisionSpecifications []*Collision,
-	reportDistances bool,
-) (Constraint, error) {
-	return newCollisionConstraint(frame, nil, observationInput, collisionSpecifications, reportDistances)
-}
-
-// newObstacleConstraint creates a constraint that will be violated if geometries constituting the given frame ever come
-// into collision with worldState geometries outside of the collisions present for the observationInput.
-// Collisions specified as collisionSpecifications will also be ignored
-// if reportDistances is false, this check will be done as fast as possible, if true maximum information will be available for debugging.
-func newObstacleConstraint(frame referenceframe.Frame,
 	fs referenceframe.FrameSystem,
-	worldState *referenceframe.WorldState,
 	observationInput map[string][]referenceframe.Input,
 	collisionSpecifications []*Collision,
 	reportDistances bool,
 ) (Constraint, error) {
-	// TODO(rb) it is bad practice to assume that the current inputs of the robot correspond to the passed in world state
-	// the state that observed the worldState should ultimately be included as part of the worldState message
-	worldState, err := worldState.ToWorldFrame(fs, observationInput)
+	// extract inputs corresponding to the frame
+	var frameInputs []referenceframe.Input
+	var err error
+	switch f := frame.(type) {
+	case *solverFrame:
+		frameInputs, err = f.mapToSlice(observationInput)
+	default:
+		frameInputs, err = referenceframe.GetFrameInputs(f, observationInput)
+	}
 	if err != nil {
 		return nil, err
 	}
-	// can use zeroth element of worldState.Obstacles because ToWorldFrame returns only one GeometriesInFrame
-	return newCollisionConstraint(frame, worldState.Obstacles[0].Geometries(), observationInput, collisionSpecifications, reportDistances)
+
+	// create robot collision entities
+	movingGeometries, err := frame.Geometries(frameInputs)
+	if err != nil && len(movingGeometries.Geometries()) == 0 {
+		return nil, err // no geometries defined for frame
+	}
+
+	// find all geometries that are not moving
+	staticGeometries, err := referenceframe.FrameSystemGeometries(fs, observationInput)
+	if err != nil {
+		return nil, err
+	}
+	for _, geometry := range movingGeometries.Geometries() {
+		staticGeometries.RemoveGeometry(geometry.Label())
+	}
+
+	movingVsMovingConstraint, err := newCollisionConstraint(
+		movingGeometries.Geometries(),
+		nil,
+		collisionSpecifications,
+		reportDistances,
+	)
+	if err != nil {
+		return nil, err
+	}
+	movingVsStaticConstraint, err := newCollisionConstraint(
```

Comment:
This doesn't make sense to put this here. The self-collision constraint is moving vs moving, and the obstacle constraint is where what you've called "moving vs static" should be. Note the old description for obstacle constraint is general about "in the solver frame vs out of it", there's no meaningful distinction between whether a static geometry comes from some other robot component or from a worldstate.

Reply:
I said the same thing but got overruled by Fahmina

Reply:
Leaving aside product management's opinions about internal implementation details, this is still the wrong way to do this.

If we want to increase visibility in "what specifically failed as a constraint", fine, then create each of self-collision, external-collision, and world-state-collision constraints separately. Let it be three differently-named constraints.

Combining the two of these as is done here is the worst possible outcome.

Reply:
Implemented this change